<a href="https://colab.research.google.com/github/jha-vineet69/Bihar-Elections-Analysis/blob/main/Bihar_Election_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <center>Bihar Assembly Election Analysis</center>
## <center>By Vineet Jha</center>
<figure>
<center>
<img src='https://drive.google.com/uc?id=1jZgHJKn11WIbOwx9NW3l-_GccBDZ9szt'/>
</center>
</figure>

# Introduction

Elections are a major event in any country and India being the biggest democracy takes it to whole another level. Bihar, the third-largest state of India by population, has one of the fiercest rivalries in RJD(Rashtriya Janata Dal) and JDU(Janata Dal United).

# Objective

While it would be interesting to visualise various metrics of Bihar Elections, the purpose of this project is to brush up on Pandas skills. I decided on a set of questions to answer using Pandas. Pandas can be used to dive deep into data and find relevant insights and help in decision making process.

# Application

### Dataset Description:
The dataset consists of 8 columns: 

1.   Election Year which in our case has values 2005-Feb, 2005-Oct, 2010 and 2015.
2.   AC Name
3.   AC No.<br>
These are the Name and Number(ranging from 1 to 243) of the Assembly constituency.
4.   Name i.e. name of the candidate contesting elections.
5.   Party - the Political party one belongs to.
6.   Votes
7.   Votes%
8.   Position - Rank obtained in the election according to Votes and Votes%




In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("https://raw.githubusercontent.com/jha-vineet69/Bihar-Elections-Analysis/main/Bihar_Elections_Data_Cleaned.csv")
df.head()

Election Year  Position  ...        AC name  AC No
0          2010         1  ...  Valmiki Nagar      1
1          2010         2  ...  Valmiki Nagar      1
2          2010         3  ...  Valmiki Nagar      1
3          2010         4  ...  Valmiki Nagar      1
4          2010         5  ...  Valmiki Nagar      1

[5 rows x 8 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12301 entries, 0 to 12300
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Election Year  12301 non-null  object
 1   Position       12301 non-null  int64 
 2   Name           12301 non-null  object
 3   Votes          12301 non-null  int64 
 4   Votes %        12301 non-null  object
 5   Party          12301 non-null  object
 6   AC name        12301 non-null  object
 7   AC No          12301 non-null  int64 
dtypes: int64(3), object(5)
memory usage: 768.9+ KB


In [4]:
df.columns

Index(['Election Year', 'Position', 'Name', 'Votes', 'Votes %', 'Party',
       'AC name', 'AC No'],
      dtype='object')

### Data Cleaning:

In [5]:
df = df.drop_duplicates()

In [6]:
df.isna().any()

Election Year    False
Position         False
Name             False
Votes            False
Votes %          False
Party            False
AC name          False
AC No            False
dtype: bool

In [7]:
df['Votes %'] = df['Votes %'].str.replace('%', '')
df.head()

Election Year  Position  ...        AC name  AC No
0          2010         1  ...  Valmiki Nagar      1
1          2010         2  ...  Valmiki Nagar      1
2          2010         3  ...  Valmiki Nagar      1
3          2010         4  ...  Valmiki Nagar      1
4          2010         5  ...  Valmiki Nagar      1

[5 rows x 8 columns]

In [8]:
df['Votes %'] = df['Votes %'].astype('float16')

In [9]:
fcols = df.select_dtypes('float').columns
icols = df.select_dtypes('integer').columns

df[fcols] = df[fcols].apply(pd.to_numeric, downcast='float')
df[icols] = df[icols].apply(pd.to_numeric, downcast='integer')
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12301 entries, 0 to 12300
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Election Year  12301 non-null  object 
 1   Position       12301 non-null  int8   
 2   Name           12301 non-null  object 
 3   Votes          12301 non-null  int32  
 4   Votes %        12301 non-null  float16
 5   Party          12301 non-null  object 
 6   AC name        12301 non-null  object 
 7   AC No          12301 non-null  int16  
dtypes: float16(1), int16(1), int32(1), int8(1), object(4)
memory usage: 588.6+ KB


In [10]:
df['AC name'].value_counts()

Buxar               84
Mahua               74
Kumhrarh            72
Kurhani             69
Riga                68
                    ..
Triveniganj (SC)     8
Katoria (ST)         8
Imamganj (SC)        8
Barachatti (SC)      6
Garkha (SC)          6
Name: AC name, Length: 282, dtype: int64

We have 243 constituencies i.e. AC No. from 1 to 243 but 282 AC names. This is due to the fact that some consistuencies were reserved for SC and ST candidates and those constituencies have SC and ST appended to their names. We can remove that as of now.

In [11]:
df['Election Year'].value_counts()

2010        3523
2015        3450
2005-Feb    3193
2005-Oct    2135
Name: Election Year, dtype: int64

In [12]:
df.replace(to_replace=' \(SC\)', value='', regex=True, inplace=True)
df.replace(to_replace=' \(ST\)', value='', regex=True, inplace=True)

In [13]:
df['AC name'].value_counts()

Buxar       84
Phulwari    79
Mahua       74
Masaurhi    73
Kumhrarh    72
            ..
Baniapur    33
Imamganj    31
Garkha      31
Katoria     31
Supaul      28
Name: AC name, Length: 243, dtype: int64

In [14]:
df[['AC name', 'AC No']].value_counts().reset_index().rename(columns={0:"Num Rows"}).sort_values(['AC No'], ascending=True)

AC name  AC No  Num Rows
131  Valmiki Nagar      1        49
130       Ramnagar      2        49
9     Narkatiaganj      3        66
80          Bagaha      4        54
106        Lauriya      5        51
..             ...    ...       ...
36     Warsaliganj    239        59
206       Sikandra    240        42
113          Jamui    241        51
32          Jhajha    242        60
120         Chakai    243        50

[243 rows x 3 columns]

#### Task 1: Determine the winning party for every election at the state level. A party which has the highest total votes is considered the winner here.

In [15]:
df_cleaned = df[df['Position'] == 1]

In [16]:
df_cleaned = df_cleaned[df_cleaned['Party'] != 'Independent']

In [17]:
df_win = df_cleaned.groupby(['Election Year','Party'])['Votes'].sum().reset_index().rename(columns={"Votes":"Total Votes"})
df_win = df_win.sort_values(['Election Year','Total Votes'], ascending=False)

In [18]:
df_win.loc[df_win.groupby(['Election Year'])['Total Votes'].idxmax()].style.hide_index()

In [19]:
df_win2 = df_cleaned.groupby(['Election Year','Party'])['Position'].count().reset_index().rename(columns={"Position":"Seats Won"})
df_win2 = df_win2.sort_values(['Election Year','Seats Won'], ascending=False)

In [20]:
df_win2.loc[df_win2.groupby(['Election Year'])['Seats Won'].idxmax()].style.hide_index()

#### Task 2: Determine the list of swing seats for the state. A swing seat is a constituency which has a different winner in every election. The elections are held in the order 2005-Feb, 2005-Oct, 2010 and 2015.

In [21]:
df_cleaned.head(12)

Election Year  Position  ...        AC name  AC No
0            2010         1  ...  Valmiki Nagar      1
14           2010         1  ...       Ramnagar      2
31           2010         1  ...   Narkatiaganj      3
46           2010         1  ...         Bagaha      4
73           2010         1  ...         Nautan      6
89           2010         1  ...      Chanpatia      7
107          2010         1  ...        Bettiah      8
143          2010         1  ...         Raxaul     10
150          2010         1  ...        Sugauli     11
160          2010         1  ...       Narkatia     12
177          2010         1  ...       Harsidhi     13
192          2010         1  ...     Govindganj     14

[12 rows x 8 columns]

In [22]:
df_swing = df_cleaned.groupby(['AC name', 'Party'])['Position'].count().reset_index().rename(columns={"Position":"Won"})

In [23]:
df_swing = df_swing.groupby(['AC name'])['Won'].max().reset_index()
df_swing

AC name  Won
0           Agiaon    2
1        Alamnagar    2
2           Alauli    2
3         Alinagar    2
4          Amarpur    2
..             ...  ...
238  Valmiki Nagar    2
239     Warisnagar    4
240    Warsaliganj    1
241      Wazirganj    2
242        Ziradei    2

[243 rows x 2 columns]

In [24]:
swing_seats = df_swing[df_swing['Won']==1]['AC name'].tolist()
print("No. of Swing States",len(swing_seats))
swing_seats

No. of Swing States 19


['Amour',
 'Bachhwara',
 'Balrampur',
 'Belsand',
 'Chakai',
 'Darauli',
 'Jamui',
 'Jhajha',
 'Kadwa',
 'Lauriya',
 'Madhubani',
 'Parsa',
 'Pipra 2',
 'Rajnagar',
 'Ramgarh',
 'Sikandra',
 'Tarari',
 'Thakurganj',
 'Warsaliganj']

#### Task 3: Determine the top 8 parties at the state-level in each election in terms of getting the most votes.

In [25]:
df_copy = df.copy(deep=True)

In [26]:
df_copy.drop(df_copy.index[df_copy['Party'] == 'Independent'], inplace=True)

In [27]:
df_votes_sum = df_copy.groupby(['Election Year','Party'])['Votes'].sum().reset_index().sort_values(['Election Year', 'Votes'], ascending=False)
df_votes_sum

Election Year                              Party    Votes
241          2015              Bharatiya Janta Party  9308015
329          2015               Rashtriya Janata Dal  6995509
279          2015                Janata Dal (United)  6416414
271          2015           Indian National Congress  2539638
293          2015               Lok Jan Shakti Party  1840834
..            ...                                ...      ...
22       2005-Feb         Indian Union Muslim League      758
0        2005-Feb             Adarsh Political Party      736
59       2005-Feb                Sanatan Samaj Party      705
12       2005-Feb           Bharat Mangalam Parishad      397
4        2005-Feb  Akhil Bharatiya Desh Bhakt Morcha      326

[368 rows x 3 columns]

In [28]:
df_top8 = df_votes_sum.loc[df_votes_sum.groupby('Election Year')['Votes'].nlargest(8).reset_index()['level_1']]
# df_top8 = df_votes_sum.groupby('Election Year').head(8).reset_index(drop=True)  #Produces same result
df_top8

Election Year                                              Party    Votes
49       2005-Feb                               Rashtriya Janata Dal  6140223
26       2005-Feb                                Janata Dal (United)  3564930
36       2005-Feb                               Lok Jan Shakti Party  3091173
13       2005-Feb                              Bharatiya Janta Party  2686290
21       2005-Feb                           Indian National Congress  1223835
10       2005-Feb                                Bahujan Samaj Party  1080745
57       2005-Feb                                    Samajwadi Party   658791
17       2005-Feb  Communist Party Of India (MARXIST-LENINIST) (L...   610345
109      2005-Oct                               Rashtriya Janata Dal  5525081
93       2005-Oct                                Janata Dal (United)  4819759
79       2005-Oct                              Bharatiya Janta Party  3686720
100      2005-Oct                               Lok Jan Shakti Party  2615901
87       2005-Oct                           Indian National Congress  1435449
75       2005-Oct                                Bahujan Samaj Party   981464
118      2005-Oct                                    Samajwadi Party   594266
84       2005-Oct  Communist Party Of India (MARXIST-LENINIST) (L...   559326
165          2010                                Janata Dal (United)  6561906
191          2010                               Rashtriya Janata Dal  5475656
149          2010                              Bharatiya Janta Party  4790436
159          2010                           Indian National Congress  2431477
175          2010                               Lok Jan Shakti Party  1957232
139          2010                                Bahujan Samaj Party   933947
183          2010                         Nationalist Congress Party   528575
154          2010  Communist Party Of India (MARXIST-LENINIST) (L...   520352
241          2015                              Bharatiya Janta Party  9308015
329          2015                               Rashtriya Janata Dal  6995509
279          2015                                Janata Dal (United)  6416414
271          2015                           Indian National Congress  2539638
293          2015                               Lok Jan Shakti Party  1840834
334          2015                          Rashtriya Lok Samta Party   976787
269          2015                   Hindustani Awam Morcha (Secular)   864856
230          2015                                Bahujan Samaj Party   788024

#### Task 4: Determine the minimum votes that each party in the top 8 of the 2015 election secured among all the past elections. Do this for all constituencies.

In [29]:
party_list = df_top8[df_top8['Election Year']=='2015']['Party'].tolist()
party_list

['Bharatiya Janta Party',
 'Rashtriya Janata Dal',
 'Janata Dal (United)',
 'Indian National Congress',
 'Lok Jan Shakti Party',
 'Rashtriya Lok Samta Party',
 'Hindustani Awam Morcha (Secular)',
 'Bahujan Samaj Party']

In [30]:
df_min = df_top8[df_top8['Party'].isin(party_list)]
df_min

Election Year                             Party    Votes
49       2005-Feb              Rashtriya Janata Dal  6140223
26       2005-Feb               Janata Dal (United)  3564930
36       2005-Feb              Lok Jan Shakti Party  3091173
13       2005-Feb             Bharatiya Janta Party  2686290
21       2005-Feb          Indian National Congress  1223835
10       2005-Feb               Bahujan Samaj Party  1080745
109      2005-Oct              Rashtriya Janata Dal  5525081
93       2005-Oct               Janata Dal (United)  4819759
79       2005-Oct             Bharatiya Janta Party  3686720
100      2005-Oct              Lok Jan Shakti Party  2615901
87       2005-Oct          Indian National Congress  1435449
75       2005-Oct               Bahujan Samaj Party   981464
165          2010               Janata Dal (United)  6561906
191          2010              Rashtriya Janata Dal  5475656
149          2010             Bharatiya Janta Party  4790436
159          2010          Indian National Congress  2431477
175          2010              Lok Jan Shakti Party  1957232
139          2010               Bahujan Samaj Party   933947
241          2015             Bharatiya Janta Party  9308015
329          2015              Rashtriya Janata Dal  6995509
279          2015               Janata Dal (United)  6416414
271          2015          Indian National Congress  2539638
293          2015              Lok Jan Shakti Party  1840834
334          2015         Rashtriya Lok Samta Party   976787
269          2015  Hindustani Awam Morcha (Secular)   864856
230          2015               Bahujan Samaj Party   788024

In [31]:
df_top8_min_votes = df_min.sort_values(['Party', 'Votes']).groupby('Party').head(1).reset_index(drop=True)

In [32]:
df_top8_min_votes

Election Year                             Party    Votes
0          2015               Bahujan Samaj Party   788024
1      2005-Feb             Bharatiya Janta Party  2686290
2          2015  Hindustani Awam Morcha (Secular)   864856
3      2005-Feb          Indian National Congress  1223835
4      2005-Feb               Janata Dal (United)  3564930
5          2015              Lok Jan Shakti Party  1840834
6          2010              Rashtriya Janata Dal  5475656
7          2015         Rashtriya Lok Samta Party   976787

#### Task 5: Determine the consistently tri-polar constituencies. i.e Constituencies in which 3 parties or more got greater than 10% of the vote shares in each election. The parties can be different for different elections.

In [33]:
# df_temp = df.drop(df.index[df['Party'] == 'Independent'])
df_temp = df.copy(deep=True)
df_temp = df_temp.sort_values(['AC name','Election Year','Votes %'], ascending=False).groupby(['AC name','Election Year']).head(3).reset_index(drop=True)

In [34]:
df_top3 = df_temp[['AC name','Election Year','Party','Votes %', 'Position']].sort_values(['AC name','Election Year']).reset_index(drop=True)
df_top3

AC name Election Year  ...    Votes %  Position
0      Agiaon      2005-Feb  ...  21.796875         1
1      Agiaon      2005-Feb  ...  18.593750         2
2      Agiaon      2005-Feb  ...  18.000000         3
3      Agiaon      2005-Oct  ...  30.203125         1
4      Agiaon      2005-Oct  ...  29.296875         2
...       ...           ...  ...        ...       ...
2911  Ziradei          2010  ...  19.000000         2
2912  Ziradei          2010  ...  11.601562         3
2913  Ziradei          2015  ...  30.000000         1
2914  Ziradei          2015  ...  25.500000         2
2915  Ziradei          2015  ...  25.406250         3

[2916 rows x 5 columns]

In [35]:
df_y = df_top3[(df_top3['Position']==3) & (df_top3['Votes %'] > 10)].groupby(['AC name'])['Election Year'].count().reset_index().rename(columns={"Election Year":"Num_Winner_Change"})
df_y

AC name  Num_Winner_Change
0           Agiaon                  4
1        Alamnagar                  3
2           Alauli                  2
3         Alinagar                  3
4          Amarpur                  2
..             ...                ...
223  Valmiki Nagar                  4
224     Warisnagar                  2
225    Warsaliganj                  1
226      Wazirganj                  3
227        Ziradei                  3

[228 rows x 2 columns]

In [36]:
tripolar_seats = df_y[df_y['Num_Winner_Change']==4]['AC name'].tolist()
tripolar_seats

['Agiaon',
 'Amnour',
 'Arwal',
 'Baisi',
 'Bakhri',
 'Balrampur',
 'Banka',
 'Chainpur',
 'Darauli',
 'Dehri',
 'Harlakhi',
 'Jagdishpur',
 'Kadwa',
 'Kanti',
 'Karakat',
 'Kargahar',
 'Katihar',
 'Lauriya',
 'Nabinagar',
 'Narkatiaganj',
 'Pranpur',
 'Ramgarh',
 'Sheikhpura',
 'Sheohar',
 'Tarari',
 'Ujiarpur',
 'Valmiki Nagar']

#### Task 6: Determine the stronghold party-seats. A stronghold seat is one that is won by the same party in all the elections.

In [37]:
df_cleaned.head()

Election Year  Position  ...        AC name  AC No
0           2010         1  ...  Valmiki Nagar      1
14          2010         1  ...       Ramnagar      2
31          2010         1  ...   Narkatiaganj      3
46          2010         1  ...         Bagaha      4
73          2010         1  ...         Nautan      6

[5 rows x 8 columns]

In [38]:
df_party_won = df_cleaned.groupby(['AC name', 'Party'])['Election Year'].count().reset_index().rename(columns={"Election Year":"Times Won"})
df_party_won

AC name                                              Party  Times Won
0       Agiaon                              Bharatiya Janta Party          1
1       Agiaon  Communist Party Of India (MARXIST-LENINIST) (L...          2
2       Agiaon                                Janata Dal (United)          1
3    Alamnagar                                Janata Dal (United)          2
4    Alamnagar                               Rashtriya Janata Dal          2
..         ...                                                ...        ...
566  Wazirganj                           Indian National Congress          1
567  Wazirganj                               Rashtriya Janata Dal          2
568    Ziradei                              Bharatiya Janta Party          1
569    Ziradei                           Communist Party Of India          2
570    Ziradei                                Janata Dal (United)          1

[571 rows x 3 columns]

In [39]:
df_times_won = df_party_won.groupby(['AC name'])['Times Won'].max().reset_index()
df_times_won

AC name  Times Won
0           Agiaon          2
1        Alamnagar          2
2           Alauli          2
3         Alinagar          2
4          Amarpur          2
..             ...        ...
238  Valmiki Nagar          2
239     Warisnagar          4
240    Warsaliganj          1
241      Wazirganj          2
242        Ziradei          2

[243 rows x 2 columns]

In [40]:
stronghold_seats = df_times_won[df_times_won['Times Won']==4]['AC name'].tolist()
stronghold_seats

['Belhar',
 'Bettiah',
 'Chanpatia',
 'Dhauraiya',
 'Gopalpur',
 'Khagaria',
 'Kurtha',
 'Maharajganj',
 'Nalanda',
 'Nirmali',
 'Parihar',
 'Rajpur',
 'Raxaul',
 'Singheshwar',
 'Siwan',
 'Warisnagar']

#### Task 7: Determine close contest constituencies. Close contest seats are those in which the top 2 finishers had a difference of less than 10% votes. Find seats in which there was a close contest in all the elections. If there is no such seat, then relax the criteria to 15%.

In [41]:
df_top2 = df.drop(df.index[df['Party'] == 'Independent'])
df_top2 = df_top2.sort_values(['AC name','Election Year','Votes %'], ascending=False).groupby(['AC name','Election Year']).head(2).reset_index(drop=True)

In [42]:
df_close_contest = df_top2[['Election Year','AC name', 'Votes %', 'Position']]

In [43]:
df_close_contest['Votes % Diff'] = df_close_contest.groupby(['Election Year','AC name'])['Votes %'].diff()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [44]:
df_close_contest = df_close_contest[abs(df_close_contest['Votes % Diff'])<10][['Election Year','AC name']]
df_close_contest = df_close_contest.groupby(['AC name'])['Election Year'].count().reset_index().rename(columns={"Election Year":"Num_CC"})

In [45]:
df_close_contest[df_close_contest['Num_CC']==4]['AC name'].tolist()

['Bahadurganj',
 'Banka',
 'Bhagalpur',
 'Dehri',
 'Goh',
 'Harlakhi',
 'Jagdishpur',
 'Karakat',
 'Kargahar',
 'Kasba',
 'Keoti',
 'Kurhani',
 'Madhubani',
 'Mohania',
 'Nathnagar',
 'Nawada',
 'Parihar',
 'Phulparas',
 'Pirpainti',
 'Pranpur',
 'Ramgarh',
 'Raxaul',
 'Sitamarhi',
 'Thakurganj',
 'Ziradei']

#### Task 8: Find the growing seats. For the top 8 parties at the state-level in the 2005-Feb Election, Determine the seats in which these parties have continuously increased their votes in the next 3 elections.

In [46]:
df_temp = df.copy(deep=True)

In [47]:
df_temp.drop(df_temp.index[df_temp['Party'] == 'Independent'], inplace=True)

In [48]:
party_list = df_top8[df_top8['Election Year']=='2005-Feb']['Party'].tolist()
party_list

['Rashtriya Janata Dal',
 'Janata Dal (United)',
 'Lok Jan Shakti Party',
 'Bharatiya Janta Party',
 'Indian National Congress',
 'Bahujan Samaj Party',
 'Samajwadi Party',
 'Communist Party Of India (MARXIST-LENINIST) (LIBERATION)']

In [49]:
df_temp = df_temp[df_temp['Party'].isin(party_list)]

In [50]:
df_temp = df_temp.groupby(['AC name', 'Party', 'Election Year'])['Votes'].sum().reset_index()

In [51]:
df_temp.sort_values(['AC name', 'Party', 'Election Year'])

AC name                  Party Election Year  Votes
0      Agiaon    Bahujan Samaj Party      2005-Feb   2693
1      Agiaon    Bahujan Samaj Party      2005-Oct   1297
2      Agiaon    Bahujan Samaj Party          2010    511
3      Agiaon    Bahujan Samaj Party          2015   1020
4      Agiaon  Bharatiya Janta Party      2005-Feb  18435
...       ...                    ...           ...    ...
4350  Ziradei   Lok Jan Shakti Party      2005-Feb  39532
4351  Ziradei   Rashtriya Janata Dal          2010  12555
4352  Ziradei        Samajwadi Party      2005-Oct   2625
4353  Ziradei        Samajwadi Party          2010    784
4354  Ziradei        Samajwadi Party          2015   3162

[4355 rows x 4 columns]

In [52]:
df_growing = df_temp.groupby(['Party', 'AC name'])['Votes'].is_monotonic_increasing.reset_index().rename(columns={"Votes":"Growing?"})
df_growing

Party        AC name  Growing?
0     Bahujan Samaj Party         Agiaon     False
1     Bahujan Samaj Party      Alamnagar     False
2     Bahujan Samaj Party         Alauli     False
3     Bahujan Samaj Party       Alinagar     False
4     Bahujan Samaj Party        Amarpur     False
...                   ...            ...       ...
1771      Samajwadi Party       Vaishali     False
1772      Samajwadi Party  Valmiki Nagar     False
1773      Samajwadi Party     Warisnagar     False
1774      Samajwadi Party      Wazirganj     False
1775      Samajwadi Party        Ziradei     False

[1776 rows x 3 columns]

In [53]:
df_growing[df_growing['Growing?']][['Party','AC name']].reset_index(drop=True)

Party      AC name
0    Bahujan Samaj Party        Banka
1    Bahujan Samaj Party    Banmankhi
2    Bahujan Samaj Party   Barachatti
3    Bahujan Samaj Party  Biharsharif
4    Bahujan Samaj Party     Chainpur
..                   ...          ...
978      Samajwadi Party  Singheshwar
979      Samajwadi Party     Sonbarsa
980      Samajwadi Party      Sugauli
981      Samajwadi Party       Tarari
982      Samajwadi Party       Tikari

[983 rows x 2 columns]

#### Task 9: Find the declining seats.

In [54]:
df_declining = df_temp.groupby(['Party', 'AC name'])['Votes'].is_monotonic_decreasing.reset_index().rename(columns={"Votes":"Declining?"})
df_declining

Party        AC name  Declining?
0     Bahujan Samaj Party         Agiaon       False
1     Bahujan Samaj Party      Alamnagar       False
2     Bahujan Samaj Party         Alauli       False
3     Bahujan Samaj Party       Alinagar       False
4     Bahujan Samaj Party        Amarpur        True
...                   ...            ...         ...
1771      Samajwadi Party       Vaishali        True
1772      Samajwadi Party  Valmiki Nagar       False
1773      Samajwadi Party     Warisnagar        True
1774      Samajwadi Party      Wazirganj        True
1775      Samajwadi Party        Ziradei       False

[1776 rows x 3 columns]

In [55]:
df_declining[df_declining['Declining?']][['Party','AC name']].reset_index(drop=True)

Party      AC name
0    Bahujan Samaj Party      Amarpur
1    Bahujan Samaj Party        Arrah
2    Bahujan Samaj Party  Bahadurganj
3    Bahujan Samaj Party   Bahadurpur
4    Bahujan Samaj Party       Barari
..                   ...          ...
693      Samajwadi Party       Tarari
694      Samajwadi Party       Tikari
695      Samajwadi Party     Vaishali
696      Samajwadi Party   Warisnagar
697      Samajwadi Party    Wazirganj

[698 rows x 2 columns]

#### Task 10: Determine the state-winner-match constituencies. Find the seats which have always elected the party that has gone on to win the state election (most votes secured at state). So if the RJD won at the state level in 2010, then seats which voted an RJD candidate as winner are considered in this category. This criteria for a seat has to be matched for each election.

In [56]:
df_x = df.copy(deep=True)

In [57]:
df_x = df_x[df_x['Position']==1][['Election Year', 'AC name', 'AC No', 'Party']]
df_x

Election Year        AC name  AC No                  Party
0              2010  Valmiki Nagar      1    Janata Dal (United)
14             2010       Ramnagar      2  Bharatiya Janta Party
31             2010   Narkatiaganj      3  Bharatiya Janta Party
46             2010         Bagaha      4    Janata Dal (United)
61             2010        Lauriya      5            Independent
...             ...            ...    ...                    ...
12242      2005-Oct    Warsaliganj    239            Independent
12261      2005-Oct       Sikandra    240  Bharatiya Janta Party
12271      2005-Oct          Jamui    241            Independent
12280      2005-Oct         Jhajha    242            Independent
12294      2005-Oct         Chakai    243  Bharatiya Janta Party

[972 rows x 4 columns]

In [58]:
df_win2

Election Year                                              Party  Seats Won
36          2015                               Rashtriya Janata Dal         80
34          2015                                Janata Dal (United)         71
30          2015                              Bharatiya Janta Party         53
33          2015                           Indian National Congress         27
31          2015  Communist Party Of India (Marxist-Leninist) (L...          3
35          2015                               Lok Jan Shakti Party          2
37          2015                          Rashtriya Lok Samta Party          2
32          2015                   Hindustani Awam Morcha (Secular)          1
26          2010                                Janata Dal (United)        115
23          2010                              Bharatiya Janta Party         91
29          2010                               Rashtriya Janata Dal         22
25          2010                           Indian National Congress          4
28          2010                               Lok Jan Shakti Party          3
24          2010                           Communist Party Of India          1
27          2010                             Jharkhand Mukti Morcha          1
18      2005-Oct                                Janata Dal (United)         88
13      2005-Oct                              Bharatiya Janta Party         55
21      2005-Oct                               Rashtriya Janata Dal         54
19      2005-Oct                               Lok Jan Shakti Party         10
17      2005-Oct                           Indian National Congress          9
16      2005-Oct  Communist Party Of India (MARXIST-LENINIST) (L...          5
12      2005-Oct                                Bahujan Samaj Party          4
14      2005-Oct                           Communist Party Of India          3
22      2005-Oct                                    Samajwadi Party          2
11      2005-Oct                                Akhil Jan Vikas Dal          1
15      2005-Oct                 Communist Party Of India (MARXIST)          1
20      2005-Oct                         Nationalist Congress Party          1
9       2005-Feb                               Rashtriya Janata Dal         75
6       2005-Feb                                Janata Dal (United)         55
1       2005-Feb                              Bharatiya Janta Party         37
7       2005-Feb                               Lok Jan Shakti Party         29
5       2005-Feb                           Indian National Congress         10
4       2005-Feb  Communist Party Of India (MARXIST-LENINIST) (L...          7
10      2005-Feb                                    Samajwadi Party          4
2       2005-Feb                           Communist Party Of India          3
8       2005-Feb                         Nationalist Congress Party          3
0       2005-Feb                                Bahujan Samaj Party          2
3       2005-Feb                 Communist Party Of India (MARXIST)          1

In [59]:
df_winners = df_win2.loc[df_win2.groupby(['Election Year'])['Seats Won'].idxmax()].reset_index(drop=True)[['Election Year', 'Party', 'Seats Won']]
df_winners

Election Year                 Party  Seats Won
0      2005-Feb  Rashtriya Janata Dal         75
1      2005-Oct   Janata Dal (United)         88
2          2010   Janata Dal (United)        115
3          2015  Rashtriya Janata Dal         80

In [60]:
class color:
   BOLD = '\033[1m'
   END = '\033[0m'

In [61]:
full_set = set((''))
for index, row in df_winners.iterrows():
  set_temp = set(df_x[(df_x['Election Year']==row[0]) & (df_x['Party']==row[1])]['AC name'])
  print(color.BOLD + f"Constituencies who voted for the winner {row[1]} in {row[0]}:" + color.END)
  print(set_temp, len(set_temp))
  if len(full_set)==0:
    full_set = set_temp
  else:
    full_set.intersection_update(set_temp)

print(color.BOLD + "Constituencies who voted for the winning party in all elections:" + color.END)
if(len(full_set)):
  print(full_set)
else:
  print("None")

Constituencies who voted for the winner Rashtriya Janata Dal in 2005-Feb:
{'Makhdumpur', 'Jokihat', 'Katoria', 'Fatuha', 'Raghopur', 'Hisua', 'Kochadhaman', 'Gaighat', 'Kesaria', 'Pranpur', 'Bihariganj', 'Gora Bauram', 'Katihar', 'Bachhwara', 'Sahebpur Kamal', 'Barhara', 'Parbatta', 'Barbigha', 'Rajauli', 'Thakurganj', 'Keoti', 'Bochaha', 'Wazirganj', 'Supaul', 'Valmiki Nagar', 'Sandesh', 'Jale', 'Kurhani', 'Pirpainti', 'Bhagalpur', 'Kutumba', 'Sikandra', 'Ghosi', 'Nokha', 'Bajpatti', 'Triveniganj', 'Benipatti', 'Kalyanpur', 'Sakra', 'Kadwa', 'Runisaidpur', 'Bakhri', 'Bihpur', 'Baruraj', 'Sursand', 'Gobindpur', 'Rupauli', 'Dinara', 'Nawada', 'Sultanganj', 'Bodh Gaya', 'Sheikhpura', 'Manjhi', 'Mohiuddinnagar', 'Kuchaikote', 'Baniapur', 'Karakat', 'Dehri', 'Dhamdaha', 'Alamnagar', 'Mahishi', 'Bhorey', 'Bahadurpur', 'Banka', 'Barhampur', 'Sasaram', 'Nabinagar', 'Sheohar', 'Imamganj', 'Buxar', 'Barauli', 'Darbhanga Rural', 'Narpatganj', 'Sugauli', 'Morwa'} 75
Constituencies who voted for t